cmd/contour/serve.go

```diff
@@ -352,6 +383,92 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
 	}
 	g.Add(debugsvc.Start)
 
+	// Leader Election starts here.
+
+	leaderOK := make(chan struct{})
+	deposed := make(chan struct{})
+	leadingDisabled := make(chan struct{})
+
+	g.Add(func(stop <-chan struct{}) error {
+		log := log.WithField("context", "leaderelect")
+		if ctx.DisableLeaderElection {
+			log.Info("Leader election disabled")
+			// if leader election is disabled, signal the gRPC goroutine
+			// to start serving and block this goroutine forever.
+			close(leaderOK)
+			<-leadingDisabled
+		}
+
+		log.Info("Started leader election")
+		log.Infof("Leader election params: %#v", ctx.LeaderElectionConfig)
+
+		// Generate the event recorder to send election events to the logs.
+		// Set up the event bits
+		eventBroadcaster := record.NewBroadcaster()
+		eventBroadcaster.StartLogging(log.Infof)
+		// Broadcast election events to the config map
+		eventBroadcaster.StartRecordingToSink(&clientcorev1.EventSinkImpl{Interface: clientcorev1.New(client.CoreV1().RESTClient()).Events("")})
+		eventsScheme := runtime.NewScheme()
+		// Need an eventsScheme
+		err := coreV1.AddToScheme(eventsScheme)
+		check(err)
+		// The recorder is what will record events about the resource lock
+		recorder := eventBroadcaster.NewRecorder(eventsScheme, coreV1.EventSource{Component: "contour"})
+
+		// Figure out the resource lock ID
+		var resourceLockID string
+		resourceLockID, isPodNameEnvSet := os.LookupEnv("POD_NAME")
+		if !isPodNameEnvSet {
+			resourceLockID = uuid.New().String()
```

Comment:
Why make this configurable? If a UUID will work when Pod name is not set, why not just use the uuid always?

Reply:
It means that you can see which pod is the leader both by inspecting the annotations on the ConfigMap and checking the Event stream on the ConfigMap. It can be difficult to tell which pod is the leader without that.